src/types/permissions.ts

```diff
@@ -42,33 +50,151 @@ export enum PermissionType {
 /**
  * Permission interfaces for different asset types
  */
-export interface FungibleAssetPermission {
-  type: PermissionType.FungibleAsset;
-  asset: string;
-  // Question: best type here?: number | string | bigint
-  balance: string;
+export class FungibleAssetPermission extends Serializable {
+  readonly type = PermissionType.FungibleAsset;
+
+  readonly asset: AccountAddress;
+
+  readonly balance: string;
+
+  constructor(asset: AccountAddress, balance: string) {
+    super();
+    this.asset = asset;
+    this.balance = balance;
+  }
+
+  static from({ asset, balance }: { asset: AccountAddress; balance: string }): FungibleAssetPermission {
+    return new FungibleAssetPermission(asset, balance);
+  }
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeStr(this.type);
+    this.asset.serialize(serializer);
+    serializer.serializeStr(this.balance);
+  }
+
+  static deserialize(deserializer: Deserializer): FungibleAssetPermission {
+    deserializer.deserializeStr() as PermissionType.FungibleAsset; // type, is this needed?;
+    const asset = AccountAddress.deserialize(deserializer);
+    const balance = deserializer.deserializeStr();
+    return new FungibleAssetPermission(asset, balance);
+  }
 }
 
-export interface GasPermission {
-  type: PermissionType.Gas;
-  amount: number;
+export class GasPermission extends Serializable {
+  readonly type = PermissionType.Gas;
+
+  readonly amount: number;
+
+  constructor(amount: number) {
+    super();
+    this.amount = amount;
+  }
+
+  static from = ({ amount }: { amount: number }): GasPermission => new GasPermission(amount);
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeStr(this.type);
+    serializer.serializeU16(this.amount);
```

Comment:
What is the appropriate type for `amount`?

Reply:
number works
gas implementation pr: https://github.com/aptos-labs/aptos-core/pull/15232/files